Dockerize Iglu Server

oguzhanunlu opened the pull request:
This PR aligns work from #11 per our docker approach (a folder with Dockerfile per version).

Iglu Registry `0.3.0` isn't published yet but this PR includes Dockerfile for that too.

Commit:
Add Dockerfile for Iglu Server (close #2)

BenFradet commented on iglu-registry/0.3.0/Dockerfile:
```diff
@@ -0,0 +1,29 @@
+
+WORKDIR $SNOWPLOW_BIN_PATH
+
+EXPOSE 8080
```
I'd rather specify the entrypoint here than in the docker compose example

oguzhanunlu replied:
Entrypoint includes `-Dconfig.file=<config_file>` which contains db configuration. Moving entrypoint here requires there is a postgres instance with that db credentials. Otherwise iglu registry will throw MissingConfiguration exception and stop. In short, running iglu-registry without postgres is useless. Should I still include entrypoint here @BenFradet ?

BenFradet replied:
yes, you need an entrypoint so you can control how the app is launched

and no you don't need to specify the config file in the entrypoint

as I said before, cf other docker images

oguzhanunlu replied:
Ah, right, I was always overriding default config.

oguzhanunlu replied:
Hey @chuwy there are two concerns we need to decide on about dockerizing iglu-registry. They came out after moving `ENTRYPOINT` from docker-compose to Dockerfile, per @BenFradet 's feedback.

1) Would that be possible to add another issue to r9 for adding `--config` CLI parameter to iglu-registry? Currently only way to provide custom config is through system property, `-Dconfig.file=<path_to_config_file>` and this config approach disables us to make real use of `CMD`  instruction when writing Dockerfile. Other dockerized apps have a `--config` CLI parameter and people can provide custom config file by mounting a config directory and providing this `--config` option.

2) If we are positive for first question, can we also add `--help` CLI parameter to iglu-registry?

@BenFradet 's suggestion is to have a uniform config approach with our apps.

alexanderdean replied:
I like this approach

chuwy replied:
@oguzhanunlu sure, it makes sense!

BenFradet commented on iglu-registry/README.md:
````diff
@@ -0,0 +1,145 @@
+The following command inserts the super API key which can be used for generating API keys with read/write privileges for a specific vendor prefix.
+
+```
+ $ docker exec iglu-registry /snowplow/bin/super_api_key_init.sh
````
why does the user need to do this? can't this be part of the container setup

oguzhanunlu replied:
This script inserts super API key, provided in `docker-compose.yml` , to `apikeys` table in postgres. I don't think iglu-registry container should run this command without being sure about postgres, which docker-compose provides.

BenFradet reviewed (commented):
It seems a lot of weight rests on the shoulders of the user , I'd rather we alleviate that weight.

You can have a look at the existing images.

Also the example might be better off in its own folder.

BenFradet commented:
also why do we need 0.2.0?

oguzhanunlu commented:
Thanks for feedback @BenFradet ! I included `0.2.0` since other apps have their previous versions dockerized. Should we keep only latest version of iglu registry?

BenFradet commented:
since we've released this project we've been incrementally adding images but not retroactively. so I don't think it's useful to add 0.2.0

oguzhanunlu commented:
Oh, alright then. I'll remove `0.2.0`.

oguzhanunlu commented:
Hey @BenFradet how does it look now?

BenFradet commented on iglu-registry/0.3.0/Dockerfile:
```diff
@@ -0,0 +1,36 @@
+
+COPY ./scripts/super_api_key_init.sh ${SNOWPLOW_BIN_PATH}
+RUN chown -R snowplow:snowplow ${SNOWPLOW_BIN_PATH}
+RUN chmod +x ${SNOWPLOW_BIN_PATH}/super_api_key_init.sh && \
```
no need for an additional layer here

oguzhanunlu replied:
Right, it has already executable bit set in git index. Will remove it.

BenFradet replied:
you should try to minimize the number of `RUN` and `COPY` that's what I meant

oguzhanunlu replied:
I see. I could chain commands.

BenFradet commented on iglu-registry/0.3.0/Dockerfile:
```diff
@@ -0,0 +1,36 @@
+    apk add postgresql-client
+
+COPY ./scripts/super_api_key_init.sh ${SNOWPLOW_BIN_PATH}
+RUN chown -R snowplow:snowplow ${SNOWPLOW_BIN_PATH}
```
this is already done above

oguzhanunlu replied:
Do you mean base image referring to "above" @BenFradet ? This is here because I noticed that, after `COPY` command, user and group permissions of `super_api_key_init.sh` is `root:root`, not `snowplow:snowplow`.

BenFradet replied:
the chown is done twice

BenFradet replied:
in your dockerfile

oguzhanunlu replied:
Oh sorry, I noticed now. Will remove first one.

BenFradet commented on iglu-registry/0.3.0/Dockerfile:
```diff
@@ -0,0 +1,36 @@
+    unzip -d ${SNOWPLOW_BIN_PATH} ${ARCHIVE} && \
+    cd /tmp && \
+    rm -rf /tmp/build && \
+    mv ${SNOWPLOW_BIN_PATH}/iglu-server-* ${SNOWPLOW_BIN_PATH}/iglu_server.jar && \
```
why the rename?

oguzhanunlu replied:
No significant reason. Will remove it.

BenFradet commented on iglu-registry/0.3.0/Dockerfile:
```diff
@@ -0,0 +1,36 @@
+COPY ./scripts/super_api_key_init.sh ${SNOWPLOW_BIN_PATH}
+RUN chown -R snowplow:snowplow ${SNOWPLOW_BIN_PATH}
+RUN chmod +x ${SNOWPLOW_BIN_PATH}/super_api_key_init.sh && \
+	sed -i -e 's/\r$//' ${SNOWPLOW_BIN_PATH}/super_api_key_init.sh
```
why is this necessary?

BenFradet commented on iglu-registry/README.md:
```diff
@@ -0,0 +1,205 @@
+
+[Iglu Registry](iglu-registry) images are based on [Snowplow base image][base-image] which leverages [the Java 8 Alpine image][alpine-image].
+
+The S3 Loader runs under [dumb-init][dumb-init] which handles reaping zombie processes
```
s3 loader

oguzhanunlu replied:
Ah, copy-pasting common stuff, will fix it.

BenFradet commented on iglu-registry/README.md:
```diff
@@ -0,0 +1,205 @@
+folder is bind mounted then ownership will be changed to the `snowplow` user.
+
+The `-XX:+UnlockExperimentalVMOptions` and `-XX:+UseCGroupMemoryLimitForHeap` JVM options will be
+automatically provided when launching the loader in order to make the JVM adhere to the memory
```
loader

BenFradet commented on iglu-registry/0.3.0/scripts/super_api_key_init.sh:
```diff
@@ -0,0 +1,7 @@
+port=$(grep port /snowplow/config/application.conf | cut -d '=' -f 2 | tr -d '"' | tr -d '\r' | tail -n1)
+dbname=$(grep dbname /snowplow/config/application.conf | cut -d '=' -f 2 | tr -d '"' | tr -d '\r')
+username=$(grep username /snowplow/config/application.conf | cut -d '=' -f 2 | tr -d '"' | tr -d '\r')
+PGPASSWORD=$(echo $pswd) psql -h $(echo $host) -p $(echo $port) -d $(echo $dbname) -U $(echo $username) -c "insert into apikeys (uid, vendor_prefix, permission, createdat) values ($(echo "'"$API_KEY"'"), '.', 'super', current_timestamp);"
```
since we're making the user launch this script, why not make the user directly launch the psql command?

Also, @jbeemster suggested that maybe we should directly specify the super api key as a secret in the config and the table bootstrapping would take care of inserting it directly in the table, which I think is a good idea.
We would need to find a way to store secrets in the config though.

Wdyt @chuwy @oguzhanunlu ?

BenFradet replied:
ah snowplow/iglu#63

BenFradet replied:
so could we remove this and make the user launch the psql command directly at least?

oguzhanunlu replied:
Sure, [this command](https://github.com/snowplow/snowplow-docker/blob/iglu-registry/iglu-registry/README.md#insert-the-super-api-key) will be changed to something like `docker exec iglu-registry psql ...` instead of `docker exec iglu-registry sh ...`.

BenFradet commented on iglu-registry/example/docker-compose.yml:
```diff
@@ -0,0 +1,30 @@
+            max-size: 1G
+        environment:
+          - API_KEY=1d9c7e70-012b-11e8-ba89-0ed5f89f718b
+    postgres:
```
Can we put postgres first since iglu-registry depends on it?

oguzhanunlu replied:
Sure.

chuwy commented:
Hey @oguzhanunlu, @BenFradet as a side-note: do you think we should add Iglu Registry to README?

chuwy commented on iglu-registry/README.md:
```diff
@@ -0,0 +1,203 @@
+
+3) Build image    
+
+    `$ docker build -t iglu-registry:0.x.0 iglu-registry/0.x.0`
```
Right command here is `docker build -t iglu-registry:0.x.0 0.x.0`

oguzhanunlu replied:
Ah, right. Will update.

chuwy commented on iglu-registry/README.md:
```diff
@@ -0,0 +1,203 @@
+  1) Running standalone image with postgres on docker host
+  2) Running with postgres in docker-compose
+
+Either way, a super API key needs to be inserted into database. `Iglu Registry` makes use of an UUID key, called super API key, to generate other API keys.
```
Unnecessary comma in `called super API key, to generate`

chuwy commented on iglu-server/README.md:
````diff
@@ -0,0 +1,203 @@
+```bash
+$ docker run \
+  -d \
+  -v ${PWD}/config:/snowplow/config \
````
We're still in `snowplow-docker/iglu-registry` (according to docs), so `snowplow-docker/iglu-registry/config` won't work. We either need to add `cd 0.x.0` somewhere or specify full paths.

chuwy replied:
I prefer first option.

BenFradet replied:
this is an example command, there is no config and there shouldn't be, except in the docker compose example of course.

chuwy replied:
Ok, makes sense. @oguzhanunlu could you please at least link `application.conf` in first paragraph of this section then.

oguzhanunlu commented:
@BenFradet I addressed your last feedback on super api key insertion and @chuwy 's feedback on both READMEs.  How does it look? Especially, updated repository README and [insert super api key method](https://github.com/snowplow/snowplow-docker/blob/iglu-registry/iglu-server/README.md#insert-the-super-api-key)

chuwy commented on iglu-server/README.md:
```diff
@@ -0,0 +1,207 @@
+
+An example docker-compose for iglu server and postgres can be found at [the docker compose example][docker-compose-example].
+
+Before running, make sure to replace the API_KEY environment variable in `iglu-server/example/docker-compose.yml` with your preferred super API key. e.g. 1d9c7e70-012b-11e8-ba89-0ed5f89f718b
```
I'm wondering if we should provide default key (this one for example). Removing one more manual step. Those who concerned about security will change it anyway. WDYT @alexanderdean, @oguzhanunlu, @BenFradet?

BenFradet replied:
yes this has already been discussed earlier in the reivew, c.f. snowplow/iglu#63

BenFradet replied:
I don't think this part is still needed anyway given the `Insert the super API key` section below

oguzhanunlu replied:
Then I'll make use of this example key as default and put a reminder in README for people concerned about security.

oguzhanunlu replied:
@BenFradet right, no need to provide that environment variable.

BenFradet commented on iglu-server/example/docker-compose.yml:
```diff
@@ -0,0 +1,30 @@
+          options:
+            max-size: 1G
+        environment:
+          - API_KEY=1d9c7e70-012b-11e8-ba89-0ed5f89f718b
```
is this still needed btw?

oguzhanunlu replied:
Again tmp commit. Needed for QA.

chuwy commented on iglu-server/README.md:
````diff
@@ -0,0 +1,207 @@
+
+```
+curl -X POST \
+  http://localhost:8080/api/auth/keygen?vendor_prefix=<a_vendor_prefix> \
````
`localhost` doesn't work due CORS. Did we change something regarding that?

oguzhanunlu replied:
I just released `rc5` of iglu server which resolve this issue.

BenFradet commented on iglu-server/0.3.0/docker-entrypoint.sh:
```diff
@@ -0,0 +1,12 @@
+# Make sure we run the server as the snowplow user
+exec su-exec snowplow:snowplow /usr/bin/java \
+  $SP_JAVA_OPTS -XX:+UnlockExperimentalVMOptions -XX:+UseCGroupMemoryLimitForHeap \
+  -jar ${SNOWPLOW_BIN_PATH}/iglu-server-0.3.0-rc3.jar "$@"
```
remember to change that to use `IGLU_SERVER_VERSION` once 0.3.0 is released :+1:

oguzhanunlu replied:
There are 2 commits and this change is inside tmp commit @BenFradet , once QA is finished, I'll remove tmp commit.

BenFradet commented:
also @oguzhanunlu don't forget to get the image into the travis matrix

alexanderdean commented:
We are calling this Iglu Server to prevent confusion...

oguzhanunlu commented:
@chuwy @BenFradet I addressed last set of feedback on both r9 PR and this PR. I published new artifact and tested both the jar itself and dockerized version. Let me know if there is something missing.

BenFradet commented on iglu-server/README.md:
```diff
@@ -0,0 +1,204 @@
+
+3) Build image    
+
+    `$ docker build -t iglu-server:0.x.0 0.x.0`
```
can you use $VERSION like what has been done in 67d96e75da1b1ecf292dc189b2411d3e2e108ac6?

BenFradet reviewed (approved):
LGTM :+1:

oguzhanunlu commented:
@BenFradet I updated example `application.conf` and tested per [new config parameter](https://github.com/snowplow/iglu/issues/362) introduced for iglu server. It solves deployment issue with Swagger UI. [Related issue](https://github.com/snowplow/iglu/issues/362#issue-312568282) has a background story if you'd like to look at.

Let me know if there is anything breaking your LGTM confirmation.

BenFradet commented:
>Let me know if there is anything breaking your LGTM confirmation.

nope all good for me, I just renamed the issue though

chuwy reviewed (approved):
Nice work, @oguzhanunlu!

alexanderdean commented:
Woohoo!

BenFradet commented:
cherry-picked into #42, closing. Thanks!
